File: challenge10swift/challenge10/challenge10/Location.swift

protocol Location {
    var name:String { get }
}

protocol Purchaseable:Location {
    var purchasePrice:GBP { get }
}

protocol Buildable:Purchaseable {
    var miniStore:DevelopmentType.BuildCostAndRent { get }
    var supermarket:DevelopmentType.BuildCostAndRent { get }
    var megaStore:DevelopmentType.BuildCostAndRent { get }
}

struct FreeParking:Location {
    let name = "Free Parking"
}

struct Go:Location {
    let name = "Go"
    let fee = GBP(100)
}

struct FactoryOrWarehouse:Purchaseable{
    let name:String
    let purchasePrice:GBP = GBP(100)
    let rent = GBP(20)
}

struct RetailSite:Buildable {
    let group:Group
    let name:String
    let purchasePrice:GBP
    let undeveloped:DevelopmentType.RentOnly
    let miniStore:DevelopmentType.BuildCostAndRent
    let supermarket:DevelopmentType.BuildCostAndRent
    let megaStore:DevelopmentType.BuildCostAndRent
}
enum Group {
    case Red
    case Green
    case Orange
}
    
struct DevelopmentType {
    struct RentOnly {let rent:GBP}
    struct BuildCostAndRent{let buildCost:GBP; let rent:GBP}
}

func == (lhs:Location, rhs:Location) -> Bool {
    return lhs.name == rhs.name
}

File: challenge12/src/main/kotlin/Location.kt
interface Location

interface Purchaseable:Location {
    val purchasePrice:GBP
}
interface Buildable:Purchaseable {
    val miniStore:DevelopmentType.BuildCostAndRent
    val supermarket:DevelopmentType.BuildCostAndRent
    val megastore:DevelopmentType.BuildCostAndRent
}

object FreeParking:Location

object Go:Location {val fee = GBP(100)}

class FactoryOrWarehouse (val name:String):Purchaseable {
    override val purchasePrice = GBP(100)
    val rent = GBP(20)
}

class RetailSite (val group:Group, val name:String, override val purchasePrice:GBP,
                  val undeveloped:DevelopmentType.RentOnly,
                  override val miniStore:DevelopmentType.BuildCostAndRent,
                  override val supermarket: DevelopmentType.BuildCostAndRent,
                  override val megastore: DevelopmentType.BuildCostAndRent
):Buildable

sealed class DevelopmentType{
    class RentOnly(val rent:GBP)
    class BuildCostAndRent(val buildingCost:GBP, val rent:GBP)
}

enum class Group {
    Red, Green, Blue, Brown, Orange, Purple
}
